File: algorithms/S/Spiral Matrix III/Spiral Matrix III.js
var spiralMatrixIII = function(rows, cols, rStart, cStart) {
    let output = [];
    let r = rStart, c = cStart;
    let right = true, down = false, left = false, up = false;
    let steps = 1, maxSteps = 2;

    while (output.length !== (rows * cols)) {
        if (r >= 0 && c >= 0 && r < rows && c < cols) output.push([r, c]);
        if (right) {
            if (steps + 1 > maxSteps) {
                steps = 2, r++;
                right = false, down = true;
            } else c++, steps++;
        }
        else if (down) {
            if (steps + 1 > maxSteps) {
                steps = 2, c--;
                down = false, left = true;
                maxSteps++;
            } else r++, steps++;
        }
        else if (left) {
            if (steps + 1 > maxSteps) {
                steps = 2, r--;
                left = false, up = true;
            } else c--, steps++;
        }
        else if (up) {
            if (steps + 1 > maxSteps) {
                steps = 2, c++;
                up = false, right = true;
                maxSteps++;
            } else r--, steps++;
        }
    }

    return output;
};

File: scripts/algorithms/S/Spiral Matrix III/Spiral Matrix III.cpp
// Runtime: 35 ms (Top 23.18%) | Memory: 12.2 MB (Top 76.14%)
class Solution {
public:
    vector<vector<int>> spiralMatrixIII(int rows, int cols, int rStart, int cStart) {

        vector<vector<int>> ans;
        ans.push_back({rStart, cStart}); //Pushing the starting point in answer
        int topRow = rStart-1; //Row above starting point
        int bottomRow = rStart+1; //Row below starting point
        int leftCol = cStart-1; //Col left to starting point
        int rightCol = cStart+1; //Col right to starting point

        while(topRow != -1 || bottomRow != rows || leftCol != -1 || rightCol != cols){ //Untill all rows and columns are exhausted

            if(rightCol != cols){ //Checking if this col is exhausted
                for(int i = topRow+1; i < bottomRow;i++) ans.push_back({i, rightCol}); //Running loop from one bottom of top row till bottom row in right col
                rightCol++; //Incrementing the col forward
            }
            //Similar things are done below with different cols and rows
            if(bottomRow != rows){
                for(int j = rightCol-1; j > leftCol; j--) ans.push_back({bottomRow, j});
                bottomRow++;
            }

            if(leftCol != -1){
                for(int i = bottomRow-1; i > topRow; i--) ans.push_back({i, leftCol});
                leftCol--;
            }

            if(topRow != -1){
                for(int j = leftCol+1; j < rightCol; j++) ans.push_back({topRow, j});
                topRow--;
            }
        }
        return ans;
    }
};